File: beefdown-rs/src/music/notes.rs
use std::collections::HashMap;

/// Convert a note name and octave to a MIDI note number (0-127)
pub fn note_to_midi(note: &str, octave: u8) -> Result<u8, String> {
    // Note name to semitone offset from C
    let note_offsets: HashMap<&str, i8> = [
        ("C", 0), ("C#", 1), ("Db", 1),
        ("D", 2), ("D#", 3), ("Eb", 3),
        ("E", 4),
        ("F", 5), ("F#", 6), ("Gb", 6),
        ("G", 7), ("G#", 8), ("Ab", 8),
        ("A", 9), ("A#", 10), ("Bb", 10),
        ("B", 11),
    ]
    .iter()
    .cloned()
    .collect();

    let offset = note_offsets
        .get(note)
        .ok_or_else(|| format!("Invalid note name: {}", note))?;

    // MIDI note number = (octave + 1) * 12 + offset
    // Octave -1 starts at MIDI note 0
    let midi_number = ((octave as i16 + 1) * 12 + *offset as i16) as u8;

    if midi_number > 127 {
        return Err(format!("Note {}{}  out of MIDI range", note, octave));
    }

    Ok(midi_number)
}

/// Parse a note string like "C4", "F#5", "Bb3" into note name and octave
pub fn parse_note(input: &str) -> Result<(&str, u8), String> {
    if input.len() < 2 {
        return Err(format!("Invalid note format: {}", input));
    }

    // Check for sharp/flat
    let (note, octave_str) = if input.len() > 2 && (input.chars().nth(1) == Some('#') || input.chars().nth(1) == Some('b')) {
        (&input[..2], &input[2..])
    } else {
        (&input[..1], &input[1..])
    };

    let octave = octave_str
        .parse::<u8>()
        .map_err(|_| format!("Invalid octave: {}", octave_str))?;

    Ok((note, octave))
}

/// Convert a note string directly to MIDI number
pub fn note_string_to_midi(input: &str) -> Result<u8, String> {
    let (note, octave) = parse_note(input)?;
    note_to_midi(note, octave)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_note_to_midi() {
        assert_eq!(note_to_midi("C", 4).unwrap(), 60); // Middle C
        assert_eq!(note_to_midi("A", 4).unwrap(), 69); // A440
        assert_eq!(note_to_midi("C", 0).unwrap(), 12); // Low C
        assert_eq!(note_to_midi("G", 9).unwrap(), 127); // Highest MIDI note
    }

    #[test]
    fn test_accidentals() {
        assert_eq!(note_to_midi("C#", 4).unwrap(), 61);
        assert_eq!(note_to_midi("F#", 4).unwrap(), 66);
        assert_eq!(note_to_midi("Db", 4).unwrap(), 61);
        assert_eq!(note_to_midi("Gb", 4).unwrap(), 66);
    }

    #[test]
    fn test_parse_note() {
        assert_eq!(parse_note("C4").unwrap(), ("C", 4));
        assert_eq!(parse_note("F#5").unwrap(), ("F#", 5));
        assert_eq!(parse_note("Bb3").unwrap(), ("Bb", 3));
    }

    #[test]
    fn test_note_string_to_midi() {
        assert_eq!(note_string_to_midi("C4").unwrap(), 60);
        assert_eq!(note_string_to_midi("A4").unwrap(), 69);
        assert_eq!(note_string_to_midi("C#5").unwrap(), 73);
    }
}

File: music/notes.go
package music

// MIDI note number functions
// Each function returns the MIDI note number for a given octave
// MIDI note numbers range from 0-127, where middle C (C4) is 60

// C returns the MIDI note number for C in the given octave
func C(octave uint8) uint8 {
	return octave*12 + 0
}

// Db returns the MIDI note number for C#/Db in the given octave
func Db(octave uint8) uint8 {
	return octave*12 + 1
}

// D returns the MIDI note number for D in the given octave
func D(octave uint8) uint8 {
	return octave*12 + 2
}

// Eb returns the MIDI note number for D#/Eb in the given octave
func Eb(octave uint8) uint8 {
	return octave*12 + 3
}

// E returns the MIDI note number for E in the given octave
func E(octave uint8) uint8 {
	return octave*12 + 4
}

// F returns the MIDI note number for F in the given octave
func F(octave uint8) uint8 {
	return octave*12 + 5
}

// Gb returns the MIDI note number for F#/Gb in the given octave
func Gb(octave uint8) uint8 {
	return octave*12 + 6
}

// G returns the MIDI note number for G in the given octave
func G(octave uint8) uint8 {
	return octave*12 + 7
}

// Ab returns the MIDI note number for G#/Ab in the given octave
func Ab(octave uint8) uint8 {
	return octave*12 + 8
}

// A returns the MIDI note number for A in the given octave
func A(octave uint8) uint8 {
	return octave*12 + 9
}

// Bb returns the MIDI note number for A#/Bb in the given octave
func Bb(octave uint8) uint8 {
	return octave*12 + 10
}

// B returns the MIDI note number for B in the given octave
func B(octave uint8) uint8 {
	return octave*12 + 11
}
